File: backend/src/middleware/auth.middleware.ts
import { Request, Response, NextFunction } from "express";
import { verifyToken, JwtPayload } from "../utils/jwt";

// Extend Express Request to carry authenticated user info
declare global {
  namespace Express {
    interface Request {
      user?: JwtPayload;
    }
  }
}

/**
 * AUTH MIDDLEWARE
 * Protects routes by requiring a valid Bearer JWT in the Authorization header.
 * Never reveals why validation failed to prevent user enumeration.
 */
export const authenticate = (
  req: Request,
  res: Response,
  next: NextFunction
): void => {
  try {
    const authHeader = req.headers.authorization;

    if (!authHeader || !authHeader.startsWith("Bearer ")) {
      res.status(401).json({ message: "Unauthorized: No token provided." });
      return;
    }

    const token = authHeader.split(" ")[1];
    const decoded = verifyToken(token);
    req.user = decoded;
    next();
  } catch {
    // Do NOT expose the specific JWT error (e.g., TokenExpiredError) — just say unauthorized.
    res.status(401).json({ message: "Unauthorized: Invalid or expired token." });
  }
};

File: AidLedger/app/middleware/auth_middleware.py
"""
JWT Authentication Middleware — Flask decorator.

Protects routes by requiring a valid Bearer JWT in the Authorization header.
Stores the decoded payload in Flask's `g` object for use in route handlers.

Security:
- Generic 401 message regardless of failure reason (prevents enumeration)
- Token payload accessible via g.current_user
"""
import logging
from functools import wraps
from flask import request, jsonify, g
from app.utils.jwt_utils import verify_token

logger = logging.getLogger(__name__)


def authenticate(f):
    """
    Decorator that enforces JWT authentication on a route.

    Usage:
        @registration_bp.post("/apply")
        @authenticate
        def apply_for_aid():
            user = g.current_user  # {'userId': ..., 'username': ..., 'role': ...}
    """
    @wraps(f)
    def decorated(*args, **kwargs):
        auth_header = request.headers.get("Authorization", "")

        if not auth_header or not auth_header.startswith("Bearer "):
            return jsonify({"message": "Unauthorized: No token provided."}), 401

        token = auth_header.split(" ", 1)[1]
        payload = verify_token(token)

        if payload is None:
            # Do NOT reveal whether the token is expired or malformed
            return jsonify({"message": "Unauthorized: Invalid or expired token."}), 401

        # Attach decoded user to Flask's request context
        g.current_user = payload
        return f(*args, **kwargs)

    return decorated
